Node removal makes the manager crash

It looks like the events callback gets called anyway even though we unregister it when disconnecting the engine:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x60 pc=0x530f2b]

goroutine 54 [running]:
runtime.panic(0x829780, 0xad6b53)
	/usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/docker/swarm/cluster.(*Engine).refreshContainer(0xc2087c91e0, 0xc2088c0440, 0x40, 0x0, 0x0, 0x0)
	/root/go/src/github.com/docker/swarm/cluster/engine.go:210 +0x1bb
github.com/docker/swarm/cluster.(*Engine).handler(0xc2087c91e0, 0xc2088c0400, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/docker/swarm/cluster/engine.go:507 +0x227
github.com/docker/swarm/cluster.*Engine.(github.com/docker/swarm/cluster.handler)·fm(0xc2088c0400, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/docker/swarm/cluster/engine.go:105 +0x58
github.com/samalba/dockerclient.(*DockerClient).getEvents(0xc2086b96c0, 0xc2087187c0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/samalba/dockerclient/dockerclient.go:295 +0x3e0
created by github.com/samalba/dockerclient.(*DockerClient).StartMonitorEvents
	/root/go/src/github.com/samalba/dockerclient/dockerclient.go:276 +0x83

goroutine 16 [chan receive, 39 minutes]:
github.com/docker/swarm/api.(*Server).ListenAndServe(0xc20872c120, 0x0, 0x0)
	/root/go/src/github.com/docker/swarm/api/server.go:113 +0x3c0
github.com/docker/swarm/cli.manage(0xc208044180)
	/root/go/src/github.com/docker/swarm/cli/manage.go:205 +0x1144
github.com/codegangsta/cli.Command.Run(0x889920, 0x6, 0x889880, 0x1, 0x8b7e50, 0x17, 0x0, 0x0, 0x0, 0x0, ...)
	/root/go/src/github.com/codegangsta/cli/command.go:113 +0xeb2
github.com/codegangsta/cli.(*App).Run(0xc208001520, 0xc20800e000, 0x8, 0x8, 0x0, 0x0)
	/root/go/src/github.com/codegangsta/cli/app.go:156 +0xaea
github.com/docker/swarm/cli.Run()
	/root/go/src/github.com/docker/swarm/cli/cli.go:133 +0xd45
main.main()
	/root/go/src/github.com/docker/swarm/main.go:13 +0x1a

goroutine 19 [finalizer wait, 42 minutes]:
runtime.park(0x412e50, 0xadb670, 0xad98c9)
	/usr/local/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0xadb670, 0xad98c9)
	/usr/local/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
	/usr/local/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
	/usr/local/go/src/pkg/runtime/proc.c:1445

goroutine 20 [select]:
github.com/docker/swarm/discovery/kv.(*Discovery).watchOnce(0xc208042380, 0x0, 0xc2087ae120, 0xc208811d40, 0xc208811da0, 0x0)
	/root/go/src/github.com/docker/swarm/discovery/kv/kv.go:71 +0x4cf
github.com/docker/swarm/discovery/kv.func·001()
	/root/go/src/github.com/docker/swarm/discovery/kv/kv.go:115 +0x20a
created by github.com/docker/swarm/discovery/kv.(*Discovery).Watch
	/root/go/src/github.com/docker/swarm/discovery/kv/kv.go:125 +0x113

goroutine 21 [select]:
github.com/docker/swarm/cluster/swarm.(*Cluster).monitorDiscovery(0xc208811ce0, 0xc208811d40, 0xc208811da0)
	/root/go/src/github.com/docker/swarm/cluster/swarm/cluster.go:218 +0x423
created by github.com/docker/swarm/cluster/swarm.NewCluster
	/root/go/src/github.com/docker/swarm/cluster/swarm/cluster.go:55 +0x356

goroutine 22 [select]:
net/http.(*persistConn).roundTrip(0xc2087c0630, 0xc2088c9a30, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1015 +0x6db
net/http.(*Transport).RoundTrip(0xc20800e600, 0xc2087c9450, 0x847fe0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:208 +0x49a
net/http.send(0xc2087c9450, 0x7f1a3713b680, 0xc20800e600, 0xc2088485d4, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:195 +0x43d
net/http.(*Client).send(0xadd1e0, 0xc2087c9450, 0x40, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:118 +0x15b
net/http.(*Client).doFollowingRedirects(0xadd1e0, 0xc2087c9450, 0x9635a8, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:343 +0x97f
net/http.(*Client).Do(0xadd1e0, 0xc2087c9450, 0xadd320, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:150 +0xa3
github.com/hashicorp/consul/api.(*Client).doRequest(0xc208811c80, 0xc20882eff0, 0x7f1a36f98cd8, 0x7f1a36f98d00, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/api.go:287 +0xdf
github.com/hashicorp/consul/api.(*KV).getInternal(0xc20803aa68, 0xc2085b2331, 0x4, 0xc20883a6f0, 0xc2085b6140, 0xadc268, 0x0, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/kv.go:104 +0x27c
github.com/hashicorp/consul/api.(*KV).List(0xc20803aa68, 0xc2085b2331, 0x4, 0xc2085b6140, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/kv.go:59 +0x137
github.com/docker/swarm/pkg/store.func·002()
	/root/go/src/github.com/docker/swarm/pkg/store/consul.go:280 +0x11b
created by github.com/docker/swarm/pkg/store.(*Consul).WatchTree
	/root/go/src/github.com/docker/swarm/pkg/store/consul.go:300 +0x172

goroutine 65 [select]:
github.com/docker/swarm/cluster.(*Engine).refreshLoop(0xc2087396c0)
	/root/go/src/github.com/docker/swarm/cluster/engine.go:282 +0x966
created by github.com/docker/swarm/cluster.(*Engine).connectClient
	/root/go/src/github.com/docker/swarm/cluster/engine.go:102 +0x172

goroutine 26 [select]:
net/http.(*persistConn).roundTrip(0xc2087c02c0, 0xc2088c9770, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1015 +0x6db
net/http.(*Transport).RoundTrip(0xc20800e600, 0xc2087c92b0, 0x95, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:208 +0x49a
net/http.send(0xc2087c92b0, 0x7f1a3713b680, 0xc20800e600, 0xc2087ee5a0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:195 +0x43d
net/http.(*Client).send(0xadd1e0, 0xc2087c92b0, 0x45, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:118 +0x15b
net/http.(*Client).doFollowingRedirects(0xadd1e0, 0xc2087c92b0, 0x9635a8, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:343 +0x97f
net/http.(*Client).Do(0xadd1e0, 0xc2087c92b0, 0xadd320, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:150 +0xa3
github.com/hashicorp/consul/api.(*Client).doRequest(0xc208811c80, 0xc20882eeb0, 0x7f1a36f92c98, 0xc2088222c0, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/api.go:287 +0xdf
github.com/hashicorp/consul/api.(*KV).getInternal(0xc20803b4e0, 0xc2086b8a21, 0x13, 0x0, 0xc208772140, 0xc208808630, 0xc208822140, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/kv.go:104 +0x27c
github.com/hashicorp/consul/api.(*KV).Get(0xc20803b4e0, 0xc2086b8a21, 0x13, 0xc208772140, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/kv.go:38 +0x97
github.com/hashicorp/consul/api.(*Lock).Lock(0xc2086e37c0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/lock.go:161 +0x4ac
github.com/docker/swarm/pkg/store.(*consulLock).Lock(0xc20803b2c0, 0x7f1a36f92f00, 0x0, 0x0)
	/root/go/src/github.com/docker/swarm/pkg/store/consul.go:324 +0x45
github.com/docker/swarm/leadership.(*Candidate).campaign(0xc2086ceb40, 0x7f1a3715d700, 0xc20803b2c0)
	/root/go/src/github.com/docker/swarm/leadership/candidate.go:90 +0x6b
created by github.com/docker/swarm/leadership.(*Candidate).RunForElection
	/root/go/src/github.com/docker/swarm/leadership/candidate.go:53 +0x14d

goroutine 17 [syscall, 39 minutes]:
runtime.goexit()
	/usr/local/go/src/pkg/runtime/proc.c:1445

goroutine 27 [chan send, 39 minutes]:
github.com/docker/swarm/pkg/store.func·001()
	/root/go/src/github.com/docker/swarm/pkg/store/consul.go:249 +0x290
created by github.com/docker/swarm/pkg/store.(*Consul).Watch
	/root/go/src/github.com/docker/swarm/pkg/store/consul.go:252 +0x172

goroutine 28 [chan receive, 39 minutes]:
github.com/docker/swarm/leadership.(*Follower).follow(0xc2084a4db0, 0xc2086cede0)
	/root/go/src/github.com/docker/swarm/leadership/follower.go:58 +0x10a
created by github.com/docker/swarm/leadership.(*Follower).FollowElection
	/root/go/src/github.com/docker/swarm/leadership/follower.go:38 +0xba

goroutine 29 [chan receive, 39 minutes]:
github.com/docker/swarm/cli.func·004()
	/root/go/src/github.com/docker/swarm/cli/manage.go:179 +0x6d
created by github.com/docker/swarm/cli.manage
	/root/go/src/github.com/docker/swarm/cli/manage.go:188 +0x10b6

goroutine 30 [chan receive, 39 minutes]:
github.com/docker/swarm/cli.func·005()
	/root/go/src/github.com/docker/swarm/cli/manage.go:192 +0x79
created by github.com/docker/swarm/cli.manage
	/root/go/src/github.com/docker/swarm/cli/manage.go:200 +0x112e

goroutine 31 [IO wait, 39 minutes]:
net.runtime_pollWait(0x7f1a3715e6f8, 0x72, 0x0)
	/tmp/makerelease290705518/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc2086dc920, 0x72, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc2086dc920, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).accept(0xc2086dc8c0, 0x963540, 0x0, 0x7f1a3713b440, 0xb)
	/usr/local/go/src/pkg/net/fd_unix.go:419 +0x343
net.(*TCPListener).AcceptTCP(0xc20803b348, 0x8, 0x0, 0x0)
	/usr/local/go/src/pkg/net/tcpsock_posix.go:234 +0x5d
net.(*TCPListener).Accept(0xc20803b348, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/tcpsock_posix.go:244 +0x4b
net/http.(*Server).Serve(0xc2086cf020, 0x7f1a3715e8a8, 0xc20803b348, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/server.go:1698 +0x91
github.com/docker/swarm/api.func·008()
	/root/go/src/github.com/docker/swarm/api/server.go:106 +0x526
created by github.com/docker/swarm/api.(*Server).ListenAndServe
	/root/go/src/github.com/docker/swarm/api/server.go:109 +0x343

goroutine 46 [IO wait]:
net.runtime_pollWait(0x7f1a3715e7a8, 0x72, 0x0)
	/tmp/makerelease290705518/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc2086dc530, 0x72, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc2086dc530, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).Read(0xc2086dc4d0, 0xc208826000, 0x1000, 0x1000, 0x0, 0x7f1a3713b440, 0xb)
	/usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
net.(*conn).Read(0xc20803b418, 0xc208826000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/net.go:122 +0xe7
net/http.noteEOFReader.Read(0x7f1a3715e948, 0xc20803b418, 0xc2087c0688, 0xc208826000, 0x1000, 0x1000, 0xaedda0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1203 +0x72
net/http.(*noteEOFReader).Read(0xc2086b9380, 0xc208826000, 0x1000, 0x1000, 0xc208809f48, 0x0, 0x0)
	<autogenerated>:124 +0xca
bufio.(*Reader).fill(0xc2086cf5c0)
	/usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
bufio.(*Reader).Peek(0xc2086cf5c0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/bufio/bufio.go:132 +0x101
net/http.(*persistConn).readLoop(0xc2087c0630)
	/usr/local/go/src/pkg/net/http/transport.go:782 +0x95
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 34 [select]:
net/http.(*persistConn).roundTrip(0xc2087c04d0, 0xc2088c9620, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1015 +0x6db
net/http.(*Transport).RoundTrip(0xc20800e600, 0xc2087c9110, 0x95, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:208 +0x49a
net/http.send(0xc2087c9110, 0x7f1a3713b680, 0xc20800e600, 0xc2087ee460, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:195 +0x43d
net/http.(*Client).send(0xadd1e0, 0xc2087c9110, 0x45, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:118 +0x15b
net/http.(*Client).doFollowingRedirects(0xadd1e0, 0xc2087c9110, 0x9635a8, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:343 +0x97f
net/http.(*Client).Do(0xadd1e0, 0xc2087c9110, 0xadd320, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:150 +0xa3
github.com/hashicorp/consul/api.(*Client).doRequest(0xc208811c80, 0xc20882edc0, 0x7f1a34657dc8, 0xc208822000, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/api.go:287 +0xdf
github.com/hashicorp/consul/api.(*KV).getInternal(0xc20803b338, 0xc2086b8ca1, 0x13, 0x0, 0xc2086e3a40, 0x41c686, 0xc2086cedd0, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/kv.go:104 +0x27c
github.com/hashicorp/consul/api.(*KV).Get(0xc20803b338, 0xc2086b8ca1, 0x13, 0xc2086e3a40, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/kv.go:38 +0x97
github.com/docker/swarm/pkg/store.func·001()
	/root/go/src/github.com/docker/swarm/pkg/store/consul.go:236 +0x109
created by github.com/docker/swarm/pkg/store.(*Consul).Watch
	/root/go/src/github.com/docker/swarm/pkg/store/consul.go:252 +0x172

goroutine 41 [select]:
net/http.(*persistConn).writeLoop(0xc2087c02c0)
	/usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 44 [IO wait]:
net.runtime_pollWait(0x7f1a3715e648, 0x72, 0x0)
	/tmp/makerelease290705518/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc2086dcf40, 0x72, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc2086dcf40, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).Read(0xc2086dcee0, 0xc208824000, 0x1000, 0x1000, 0x0, 0x7f1a3713b440, 0xb)
	/usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
net.(*conn).Read(0xc20803b410, 0xc208824000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/net.go:122 +0xe7
net/http.noteEOFReader.Read(0x7f1a3715e948, 0xc20803b410, 0xc2087c05d8, 0xc208824000, 0x1000, 0x1000, 0xaedda0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1203 +0x72
net/http.(*noteEOFReader).Read(0xc2086b9320, 0xc208824000, 0x1000, 0x1000, 0xc208809f48, 0x0, 0x0)
	<autogenerated>:124 +0xca
bufio.(*Reader).fill(0xc2086cf500)
	/usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
bufio.(*Reader).Peek(0xc2086cf500, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/bufio/bufio.go:132 +0x101
net/http.(*persistConn).readLoop(0xc2087c0580)
	/usr/local/go/src/pkg/net/http/transport.go:782 +0x95
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 42 [IO wait]:
net.runtime_pollWait(0x7f1a3715e598, 0x72, 0x0)
	/tmp/makerelease290705518/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc2086dcfb0, 0x72, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc2086dcfb0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).Read(0xc2086dcf50, 0xc208744000, 0x1000, 0x1000, 0x0, 0x7f1a3713b440, 0xb)
	/usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
net.(*conn).Read(0xc20803b408, 0xc208744000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/net.go:122 +0xe7
net/http.noteEOFReader.Read(0x7f1a3715e948, 0xc20803b408, 0xc2087c0528, 0xc208744000, 0x1000, 0x1000, 0xaedda0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1203 +0x72
net/http.(*noteEOFReader).Read(0xc2086b92c0, 0xc208744000, 0x1000, 0x1000, 0xc208809f50, 0x0, 0x0)
	<autogenerated>:124 +0xca
bufio.(*Reader).fill(0xc2086cf440)
	/usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
bufio.(*Reader).Peek(0xc2086cf440, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/bufio/bufio.go:132 +0x101
net/http.(*persistConn).readLoop(0xc2087c04d0)
	/usr/local/go/src/pkg/net/http/transport.go:782 +0x95
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 40 [IO wait]:
net.runtime_pollWait(0x7f1a3715e4e8, 0x72, 0x0)
	/tmp/makerelease290705518/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc2086dd020, 0x72, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc2086dd020, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).Read(0xc2086dcfc0, 0xc20806e000, 0x1000, 0x1000, 0x0, 0x7f1a3713b440, 0xb)
	/usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
net.(*conn).Read(0xc20803b3f8, 0xc20806e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/net.go:122 +0xe7
net/http.noteEOFReader.Read(0x7f1a3715e948, 0xc20803b3f8, 0xc2087c0318, 0xc20806e000, 0x1000, 0x1000, 0xaedda0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1203 +0x72
net/http.(*noteEOFReader).Read(0xc2086b91e0, 0xc20806e000, 0x1000, 0x1000, 0xc208809f58, 0x0, 0x0)
	<autogenerated>:124 +0xca
bufio.(*Reader).fill(0xc2086cf380)
	/usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
bufio.(*Reader).Peek(0xc2086cf380, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/bufio/bufio.go:132 +0x101
net/http.(*persistConn).readLoop(0xc2087c02c0)
	/usr/local/go/src/pkg/net/http/transport.go:782 +0x95
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 43 [select]:
net/http.(*persistConn).writeLoop(0xc2087c04d0)
	/usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 45 [select]:
net/http.(*persistConn).writeLoop(0xc2087c0580)
	/usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 47 [select]:
net/http.(*persistConn).writeLoop(0xc2087c0630)
	/usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 52 [select]:
github.com/hashicorp/consul/api.(*Session).RenewPeriodic(0xc20803b4d8, 0x87cec0, 0x3, 0xc208806510, 0x24, 0x0, 0xc2086cfc80, 0x0, 0x0)
	/root/go/src/github.com/hashicorp/consul/api/session.go:159 +0x23d
created by github.com/hashicorp/consul/api.(*Lock).Lock
	/root/go/src/github.com/hashicorp/consul/api/lock.go:134 +0x361

goroutine 50 [runnable]:
net/http.(*persistConn).readLoop(0xc2087c06e0)
	/usr/local/go/src/pkg/net/http/transport.go:868 +0x829
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 51 [select, 39 minutes]:
net/http.(*persistConn).writeLoop(0xc2087c06e0)
	/usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 60 [select, 39 minutes]:
net/http.(*persistConn).writeLoop(0xc2087c0b00)
	/usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 59 [IO wait, 39 minutes]:
net.runtime_pollWait(0x7f1a3715e388, 0x72, 0x0)
	/tmp/makerelease290705518/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc208707020, 0x72, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc208707020, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).Read(0xc208706fc0, 0xc2085cb000, 0x1000, 0x1000, 0x0, 0x7f1a3713b440, 0xb)
	/usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
net.(*conn).Read(0xc20803b750, 0xc2085cb000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/net.go:122 +0xe7
net/http.noteEOFReader.Read(0x7f1a3715e948, 0xc20803b750, 0xc2087c0b58, 0xc2085cb000, 0x1000, 0x1000, 0xaedda0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1203 +0x72
net/http.(*noteEOFReader).Read(0xc20867b2c0, 0xc2085cb000, 0x1000, 0x1000, 0xc20861b770, 0x0, 0x0)
	<autogenerated>:124 +0xca
bufio.(*Reader).fill(0xc20868fc80)
	/usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
bufio.(*Reader).Peek(0xc20868fc80, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/bufio/bufio.go:132 +0x101
net/http.(*persistConn).readLoop(0xc2087c0b00)
	/usr/local/go/src/pkg/net/http/transport.go:782 +0x95
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 63 [select, 39 minutes]:
net/http.(*persistConn).readLoop(0xc2087c0bb0)
	/usr/local/go/src/pkg/net/http/transport.go:868 +0x829
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 64 [select, 39 minutes]:
net/http.(*persistConn).writeLoop(0xc2087c0bb0)
	/usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 66 [select]:
net/http.(*persistConn).roundTrip(0xc2087c0d10, 0xc2088c9cf0, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1015 +0x6db
net/http.(*Transport).RoundTrip(0xc20889a280, 0xc2087c96c0, 0xe0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:208 +0x49a
net/http.send(0xc2087c96c0, 0x7f1a3713b680, 0xc20889a280, 0xc20805e1c0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:195 +0x43d
net/http.(*Client).send(0xc2088995c0, 0xc2087c96c0, 0x8c, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:118 +0x15b
net/http.(*Client).doFollowingRedirects(0xc2088995c0, 0xc2087c96c0, 0x9635a8, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:343 +0x97f
net/http.(*Client).Do(0xc2088995c0, 0xc2087c96c0, 0xc, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/client.go:150 +0xa3
github.com/samalba/dockerclient.(*DockerClient).doStreamRequest(0xc2086638a0, 0x87f960, 0x3, 0xc20877e280, 0x77, 0x7f1a3715d6b0, 0xc208848bd0, 0x0, 0x0, 0x0, ...)
	/root/go/src/github.com/samalba/dockerclient/dockerclient.go:97 +0x3ae
github.com/samalba/dockerclient.(*DockerClient).doRequest(0xc2086638a0, 0x87f960, 0x3, 0xc20877e280, 0x77, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/root/go/src/github.com/samalba/dockerclient/dockerclient.go:70 +0x165
github.com/samalba/dockerclient.(*DockerClient).ListContainers(0xc2086638a0, 0x1, 0xc20882f130, 0x4b, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/samalba/dockerclient/dockerclient.go:148 +0x28f
github.com/docker/swarm/cluster.(*Engine).refreshContainer(0xc2087396c0, 0xc2088c03c0, 0x40, 0x0, 0x0, 0x0)
	/root/go/src/github.com/docker/swarm/cluster/engine.go:210 +0x1c1
github.com/docker/swarm/cluster.(*Engine).handler(0xc2087396c0, 0xc2088c0380, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/docker/swarm/cluster/engine.go:507 +0x227
github.com/docker/swarm/cluster.*Engine.(github.com/docker/swarm/cluster.handler)·fm(0xc2088c0380, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/docker/swarm/cluster/engine.go:105 +0x58
github.com/samalba/dockerclient.(*DockerClient).getEvents(0xc2086638a0, 0xc2088ab740, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/samalba/dockerclient/dockerclient.go:295 +0x3e0
created by github.com/samalba/dockerclient.(*DockerClient).StartMonitorEvents
	/root/go/src/github.com/samalba/dockerclient/dockerclient.go:276 +0x83

goroutine 68 [IO wait]:
net.runtime_pollWait(0x7f1a3715e228, 0x72, 0x0)
	/tmp/makerelease290705518/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc2088c73a0, 0x72, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc2088c73a0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).Read(0xc2088c7340, 0xc2088d0000, 0x1000, 0x1000, 0x0, 0x7f1a3713b440, 0xb)
	/usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
net.(*conn).Read(0xc20803bae0, 0xc2088d0000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/net.go:122 +0xe7
net/http.noteEOFReader.Read(0x7f1a3715e948, 0xc20803bae0, 0xc2087c0d68, 0xc2088d0000, 0x1000, 0x1000, 0xaedda0, 0x0, 0x0)
	/usr/local/go/src/pkg/net/http/transport.go:1203 +0x72
net/http.(*noteEOFReader).Read(0xc208626500, 0xc2088d0000, 0x1000, 0x1000, 0xc2085e1140, 0x0, 0x0)
	<autogenerated>:124 +0xca
bufio.(*Reader).fill(0xc20860cfc0)
	/usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
bufio.(*Reader).Peek(0xc20860cfc0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/pkg/bufio/bufio.go:132 +0x101
net/http.(*persistConn).readLoop(0xc2087c0d10)
	/usr/local/go/src/pkg/net/http/transport.go:782 +0x95
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 69 [select]:
net/http.(*persistConn).writeLoop(0xc2087c0d10)
	/usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
	/usr/local/go/src/pkg/net/http/transport.go:601 +0x957
exit status 2
```

--- a/cluster/engine.go
+++ b/cluster/engine.go
@@ -112,7 +112,8 @@ func (e *Engine) ConnectWithClient(client dockerclient.Client) error {
 // Disconnect will stop all monitoring of the engine.
 // The Engine object cannot be further used without reconnecting it first.
 func (e *Engine) Disconnect() {
-	close(e.stopCh)
+	// do not close the chan, so it wait until the refreshLoop goroutine stops
+	e.stopCh <- struct{}{}
 	e.client.StopAllMonitorEvents()
 	e.client = nil
 	e.emitEvent("engine_disconnect")

